### PR TITLE
Theming

### LINKED ISSUE #657: (Feature request) Themes
Add UI color accent customization and a way to share themes with users.

### PR DESCRIPTION
**Pull request by yougotwill**
Closes https://github.com/oxen-io/session-desktop-temp/issues/548

### CODE REVIEW

File: ts/components/conversation/SessionRightPanel.tsx [hunk 1]
```diff
@@ -141,6 +142,25 @@ const HeaderItem = () => {
   );
 };
 
+const StyledLeaveButton = styled.div`
+  width: 100%;
+  .session-button {
```
Comment (Bilb) [inline ts/components/conversation/SessionRightPanel.tsx:147]: someday, we will get rid of the .session-button class altogether. But not today

File: ts/components/conversation/message/reactions/ReactionPopup.tsx [hunk 2]
```diff
@@ -96,7 +96,7 @@ const generateContactsString = async (
 };
 
 const Contacts = (contacts: Array<string>, count: number) => {
-  const darkMode = useSelector(getTheme) === 'dark';
+  const darkMode = useSelector(getTheme).includes('dark');
```
Comment (Bilb) [inline ts/components/conversation/message/reactions/ReactionPopup.tsx:99]: same here, to be replaced with a useSelector(isDarkTheme) or something

File: ts/components/dialog/BanOrUnbanUserDialog.tsx [hunk 3]
```diff
@@ -65,6 +67,7 @@ export const BanOrUnBanUserDialog = (props: {
   const { i18n } = window;
   const isBan = banType === 'ban';
   const dispatch = useDispatch();
+  const darkMode = useSelector(getTheme).includes('dark');
```
Comment (Bilb) [inline ts/components/dialog/BanOrUnbanUserDialog.tsx:70]: same

File: ts/components/dialog/ModeratorsAddDialog.tsx [hunk 4]
```diff
@@ -18,6 +20,7 @@ export const AddModeratorsDialog = (props: Props) => {
   const { conversationId } = props;
 
   const dispatch = useDispatch();
+  const darkMode = useSelector(getTheme).includes('dark');
```
Comment (Bilb) [inline ts/components/dialog/ModeratorsAddDialog.tsx:23]: same

File: ts/state/ducks/theme.tsx [hunk 5]
```diff
@@ -9,7 +9,7 @@ export const applyTheme = (theme: ThemeStateType) => {
   };
 };
 
-export const initialThemeState: ThemeStateType = 'light';
+export const initialThemeState: ThemeStateType = 'classic-dark';
```
Comment (Bilb) [inline ts/state/ducks/theme.tsx:12]: just want to make sure we double check that we want this to be the default

File: ts/themes/SessionTheme.tsx [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+import { createGlobalStyle } from 'styled-components';
+import { classicDark } from './';
+import { declareCSSVariables, THEME_GLOBALS } from './globals';
+
+// Defaults to Classic Dark theme
+export const SessionGlobalStyles = createGlobalStyle`
+  html {
+    ${declareCSSVariables(THEME_GLOBALS)}
+    ${declareCSSVariables(classicDark)}
```
Comment (Bilb) [inline ts/themes/SessionTheme.tsx:11]: nice!

File: ts/themes/SessionTheme.tsx [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+
+import { createGlobalStyle } from 'styled-components';
+import { classicDark } from './';
+import { declareCSSVariables, THEME_GLOBALS } from './globals';
+
+// Defaults to Classic Dark theme
+export const SessionGlobalStyles = createGlobalStyle`
```
Comment (Bilb) [inline ts/themes/SessionTheme.tsx:8]: does this need to be exported?

File: ts/themes/switchPrimaryColor.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { find } from 'lodash';
+import { Dispatch } from '@reduxjs/toolkit';
+import { applyPrimaryColor } from '../state/ducks/primaryColor';
+import { COLORS, getPrimaryColors, PrimaryColorStateType } from './constants/colors';
+
+export function findPrimaryColorId(hexCode: string): PrimaryColorStateType | undefined {
+  const primaryColors = getPrimaryColors();
+  return find(primaryColors, { color: hexCode })?.id;
+}
+
```
Comment (Bilb) [inline ts/themes/switchPrimaryColor.tsx:10]: can you point me to which files calls this function?

File: ts/themes/switchPrimaryColor.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+export async function switchPrimaryColorTo(color: PrimaryColorStateType, dispatch?: Dispatch) {
+  if (window.Events) {
+    await window.Events.setPrimaryColorSetting(color);
+  }
+
+  document.documentElement.style.setProperty(
+    '--primary-color',
+    (COLORS.PRIMARY as any)[`${color.toUpperCase()}`]
```
Comment (Bilb) [inline ts/themes/switchPrimaryColor.tsx:18]: would be nice to get rid of the any here

File: ts/state/ducks/section.tsx [hunk 10]
```diff
@@ -10,7 +10,7 @@ export enum SectionType {
   Profile,
   Message,
   Settings,
-  Moon,
+  ColorMode,
```
Comment (Bilb) [inline ts/state/ducks/section.tsx:13]: not sure about the name of this

File: ts/components/dialog/SessionModal.tsx [hunk 11]
```diff
@@ -27,6 +27,7 @@ interface State {
   isVisible: boolean;
 }
 
+// NOTE This is currently unused.
```
Comment (Bilb) [inline ts/components/dialog/SessionModal.tsx:30]: should we keep it then?

File: ts/components/leftpane/ActionsPanel.tsx [hunk 12]
```diff
@@ -45,26 +45,37 @@ import { ipcRenderer } from 'electron';
...
-    } else if (type === SectionType.Moon) {
-      const currentTheme = window.Events.getThemeSetting();
-      const newTheme = currentTheme === 'dark' ? 'light' : 'dark';
-
-      await switchThemeTo(newTheme, dispatch);
+    } else if (type === SectionType.ColorMode) {
+      const currentTheme = String(window.Events.getThemeSetting());
+      const newTheme = (currentTheme.includes('light')
+        ? currentTheme.replace('light', 'dark')
+        : currentTheme.replace('dark', 'light')) as ThemeStateType;
```
Comment (Bilb) [inline ts/components/leftpane/ActionsPanel.tsx:70]: same, create a selector useSelector(isDarkTheme)

File: ts/components/leftpane/ActionsPanel.tsx [hunk 13]
```diff
@@ -123,13 +132,13 @@ const Section = (props: { type: SectionType }) => {
           id={'onion-path-indicator-led-id'}
         />
       );
+    case SectionType.ColorMode:
     default:
       return (
         <SessionIconButton
           iconSize="medium"
-          iconType={'moon'}
+          iconType={theme.includes('light') ? 'sun' : 'moon'}
```
Comment (Bilb) [inline ts/components/leftpane/ActionsPanel.tsx:140]: same isDarkTheme selector

File: stylesheets/_quote.scss [hunk 14]
```diff
@@ -10,273 +10,229 @@
...
+    font-size: var(--font-size-sm);
+    line-height: 18px;
+
+    color: var(--message-bubbles-received-text-color);
+    border-color: var(--message-bubbles-received-text-color);
   }
-}
+  &__primary__author {
+    font-size: var(--font-size-sm);
+    font-weight: bold;
```
Comment (Bilb) [inline stylesheets/_quote.scss:33]: next to do will be to move all of those lines to  styled components too

Comment (Bilb) [reply stylesheets/_quote.scss:33]: but not in that PR

**Review by Bilb — CHANGES_REQUESTED** [review]

File: ts/components/dialog/BanOrUnbanUserDialog.tsx (hunk 3 above)
Comment (yougotwill) [reply ts/components/dialog/BanOrUnbanUserDialog.tsx:70]: 👍

File: ts/components/dialog/ModeratorsAddDialog.tsx (hunk 4 above)
Comment (yougotwill) [reply ts/components/dialog/ModeratorsAddDialog.tsx:23]: 👍

File: ts/components/dialog/SessionModal.tsx (hunk 11 above)
Comment (yougotwill) [reply ts/components/dialog/SessionModal.tsx:30]: So I know that iOS has a caption editor and I suspect android does too. So at some point in the future we may need this in order to be "consistent with other platforms".

File: ts/components/leftpane/ActionsPanel.tsx (hunk 12 above)
Comment (yougotwill) [reply ts/components/leftpane/ActionsPanel.tsx:70]: 👍

File: ts/components/leftpane/ActionsPanel.tsx (hunk 13 above)
Comment (yougotwill) [reply ts/components/leftpane/ActionsPanel.tsx:140]: 👍

File: ts/state/ducks/section.tsx (hunk 10 above)
Comment (yougotwill) [reply ts/state/ducks/section.tsx:13]: This is what it is called in the Session Components Figma. I'm also iffy on it but we would need to discuss with Morgan and Connor

File: ts/state/ducks/theme.tsx (hunk 5 above)
Comment (yougotwill) [reply ts/state/ducks/theme.tsx:12]: Answered above 👍

File: ts/themes/SessionTheme.tsx (hunk 7 above)
Comment (yougotwill) [reply ts/themes/SessionTheme.tsx:8]: Good catch!

File: ts/themes/SessionTheme.tsx (hunk 6 above)
Comment (yougotwill) [reply ts/themes/SessionTheme.tsx:11]: Beautiful isn't it! 🔥

File: ts/themes/switchPrimaryColor.tsx (hunk 8 above)
Comment (yougotwill) [reply ts/themes/switchPrimaryColor.tsx:10]: It's used in`ts/themes/switchTheme.tsx`

File: ts/themes/switchPrimaryColor.tsx (hunk 9 above)
Comment (yougotwill) [reply ts/themes/switchPrimaryColor.tsx:18]: Fixed 👍

**Review by yougotwill — COMMENTED** [review]
Responded to feedback

File: ts/components/conversation/SessionLastSeenIndicator.tsx [hunk 15]
```diff
@@ -25,21 +15,40 @@ const LastSeenBar = styled.div`
...
+    color: ${props => (props.darkMode ? 'var(--primary-color)' : 'var(--text-primary-color)')};
+  }
 `;
 
 export const SessionLastSeenIndicator = (props: {
   messageId: string;
   didScroll: boolean;
   setDidScroll: (scroll: boolean) => void;
 }) => {
+  const darkMode = useSelector(getTheme).includes('dark');
```
[the remark this answers is not in the document]
Comment (Bilb) [reply ts/components/conversation/SessionLastSeenIndicator.tsx:51]: yes

File: ts/state/ducks/section.tsx (hunk 10 above)
Comment (Bilb) [reply ts/state/ducks/section.tsx:13]: :vomiting_face:

File: ts/components/dialog/SessionModal.tsx (hunk 11 above)
Comment (Bilb) [reply ts/components/dialog/SessionModal.tsx:30]: kill it !

File: about_preload.js [hunk 16]
```diff
@@ -8,13 +8,25 @@ const config = url.parse(window.location.toString(), true).query;
...
+const { AboutView } = require('./ts/components/AboutView');
 
-window.i18n = i18n.setupi18n(locale, localeMessages);
+window.Signal = {
+  Components: {
+    AboutView,
+  },
+};
+
+window.closeAbout = () => ipcRenderer.send('close-about');
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply about_preload.js:30]: Okay 👍

File: ts/components/conversation/SessionEmojiPanel.tsx [hunk 17]
```diff
@@ -95,44 +98,52 @@ const pickerProps: FixedPickerProps = {
...
+      default:
+        setPanelBackgroundRGB(hexColorToRGB(THEMES.CLASSIC_DARK.COLOR1));
+        setPanelTextRGB(hexColorToRGB(THEMES.CLASSIC_DARK.COLOR6));
     }
-
-    return () => {
-      isCancelled = true;
-    };
-  }, [data, pickerProps]);
+  }, [theme]);
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/components/conversation/SessionEmojiPanel.tsx]: Fixed 👍

File: ts/components/dialog/SessionModal.tsx (hunk 11 above)
Comment (yougotwill) [reply ts/components/dialog/SessionModal.tsx:30]: it rose back from the dead! 🧟

**Review by Bilb — APPROVED** [review]

**Pull request merged**